src/caching_context.jl

```diff
@@ -0,0 +1,110 @@
+"""
+    CachedHash(context)
+
+Setup a hash context that includes a cache of hash results. It stores the result of hashing
+types and large values. Calling the same cached hash context will re-use the this cache,
+possibly improving performance. If you do not pass a `CachedHash` to `stalbe_hash` it sets
+up its own internal cache to improve performance for repeated hashes of the same type or
+large value.
+
+## See Also
+
+[`stable_hash`](@ref)
+"""
+struct CachedHash{T}
+    parent::T
+    type_cache::IdDict{Type,Vector{UInt8}}
+    value_cache::IdDict{Any,Vector{UInt8}}
```

Comment:
I think these should be weak-key dictionaries so they don't keep references to the keys that prevents GC. Unfortunately there's not a super nice way to get one; I think we should copy the version from https://github.com/Nemocas/AbstractAlgebra.jl/pull/1419 and set it up as its own package...